test/token/ERC20Rebasable.unit.test.ts

```diff
@@ -0,0 +1,983 @@
+    const rebasableTokenImpl = await new ERC20Rebasable__factory(deployer).deploy(
+      "name",
+      "",
+      10,
+      wrappedTokenStub.address,
+      tokenRateOracleStub.address,
+      owner.address
+    );
+    await assert.revertsWith(
+      rebasableTokenImpl.initialize("New Name", ""),
+      "ErrorNameAlreadySet()"
+    );
+  })
+
+  .test("initialize() :: symbol already set", async (ctx) => {
+    const { deployer, owner } = ctx.accounts;
+
+    // deploy new implementation
+    const wrappedTokenStub = await new ERC20Stub__factory(deployer).deploy(); 
+    const tokenRateOracleStub = await new TokenRateOracleStub__factory(deployer).deploy(); 
+    const rebasableTokenImpl = await new ERC20Rebasable__factory(deployer).deploy(
+      "",
+      "symbol",
+      10,
+      wrappedTokenStub.address,
+      tokenRateOracleStub.address,
+      owner.address
+    );
+    await assert.revertsWith(
+      rebasableTokenImpl.initialize("", "New Symbol"),
+      "ErrorSymbolAlreadySet()"
+    );
+  })
+
+  .test("decimals", async (ctx) =>
+    assert.equal(await ctx.contracts.rebasableProxied.decimals(), ctx.constants.decimalsToSet)
+  )
+
+  .test("totalShares", async (ctx) => {
+    const { premintShares } = ctx.constants;
+    assert.equalBN(await ctx.contracts.rebasableProxied.getTotalShares(), premintShares);
+  })
+
+  .test("wrap(0)", async (ctx) => {
+    const { rebasableProxied } = ctx.contracts;
+    const { user1 } = ctx.accounts;
+    await assert.revertsWith(rebasableProxied.connect(user1).wrap(0), "ErrorZeroSharesWrap()");
+  })
+
+  .test("unwrap(0)", async (ctx) => {
+    const { rebasableProxied } = ctx.contracts;
+    const { user1 } = ctx.accounts;
+    await assert.revertsWith(rebasableProxied.connect(user1).unwrap(0), "ErrorZeroTokensUnwrap()");
+  })
+
+  .test("wrap() positive scenario", async (ctx) => {
+    
+    const { rebasableProxied, wrappedTokenStub } = ctx.contracts;
+    const {user1, user2 } = ctx.accounts;
+    const { rate, decimals, premintShares } = ctx.constants;
+
+    const totalSupply = rate.mul(premintShares).div(decimals);
+
+    assert.equalBN(await rebasableProxied.getTotalShares(), premintShares);
+    assert.equalBN(await rebasableProxied.totalSupply(), totalSupply);
+
+    // user1
+    assert.equalBN(await rebasableProxied.sharesOf(user1.address), 0);
+    assert.equalBN(await rebasableProxied.balanceOf(user1.address), 0);
+
+    const user1Shares = wei`100 ether`;
+    const user1Tokens = rate.mul(user1Shares).div(decimals);
+
+    assert.equalBN(await rebasableProxied.connect(user1).callStatic.wrap(user1Shares), user1Tokens);
+    const tx = await rebasableProxied.connect(user1).wrap(user1Shares);
+
+    assert.equalBN(await rebasableProxied.sharesOf(user1.address), user1Shares);
+    assert.equalBN(await rebasableProxied.balanceOf(user1.address), user1Tokens);
+
+    assert.equal(await wrappedTokenStub.transferFromAddress(), user1.address);
```

Comment:
I don't think we need stubs like this here and there. Can we go without them?

Especially this line checks very specific logic of the stub but a real contract. I'm not sure this is useful: think it is more likely to cause confusion

Reply:
After offline discussion we came up to the idea to leave some meaningful unit test but support them with tests where real contracts implementation is tested: RebasableToken + NonRebasableToken + Oracle.